src/Interpreters/DeadLetterQueue.h

```diff
@@ -0,0 +1,59 @@
+#pragma once
+#include <variant>
+#include <Interpreters/SystemLog.h>
+#include <Core/NamesAndTypes.h>
+#include <Core/NamesAndAliases.h>
+#include <Storages/ColumnsDescription.h>
+
+namespace DB
+{
+
+struct DeadLetterQueueElement
+{
+    enum class StreamType : int8_t
+    {
+        Kafka = 1,
+        RabbitMQ = 2,
+    };
+
+    StreamType stream_type;
+    UInt64 event_time{};
+    Decimal64 event_time_microseconds{};
+
+    String database_name;
+    String table_name;
+    String raw_message;
+    String error;
+
+    /// Engine specific details
+    struct KafkaDetails
+    {
+        String topic_name;
+        Int64 partition;
+        Int64 offset;
+    };
+    struct RabbitMQDetails
```

Comment:
Maybe also include the message key, which is in the `_key` virtual column is storage kafka.

Reply:
Because of an unclear reason _key is not populated in case of an error.